lib/FrontendTool/ScanDependencies.cpp

```diff
@@ -32,12 +32,101 @@
 #include "llvm/ADT/StringRef.h"
 #include "llvm/ADT/StringSet.h"
 #include "llvm/Support/FileSystem.h"
+#include "llvm/Support/StringSaver.h"
+#include "llvm/Support/YAMLTraits.h"
+#include "llvm/Support/YAMLParser.h"
 #include <set>
 
 using namespace swift;
+using namespace llvm::yaml;
 
 namespace {
+struct BatchScanInput {
+  StringRef moduleName;
+  StringRef arguments;
+  StringRef outputPath;
+};
+
+static std::string getScalaNodeText(Node *N) {
+  SmallString<32> Buffer;
+  return cast<ScalarNode>(N)->getValue(Buffer).str();
+}
+
+/// Parse an entry like this, where the "platforms" key-value pair is optional:
+///  {
+///     "module": "Foo.pcm",
```

Comment:
Is the only reason we include a file extension here to distinguish whether or not that's a Swift or Clang module? It seems that for either scanner we then strip it away. The file extension is kind of a downstream implementation detail here; sure, once we build them they end up being a `.swiftmodule` or a `.pcm`, but at the scanning stage we have neither and all we have to search by is a module name.

Do you think it would be cleaner to have the JSON key indicate what kind of module it is and leave module name value just have the name the scanners can use directly?

Reply:
Yeah, I think adding a more explicit key is totally reasonable. Should we add a boolean value key `isSwiftModule`?

Reply:
Yeah, either an additional boolean value, or have the module name key itself be either `clangModuleName` or `swiftModuleName`, kinda like this:  
```swift 
enum BatchScanModuleInfo {
  case swift(SwiftScanModuleInfo)
  case clang(ClangScanModuleInfo)
}
struct ClangScanModuleInfo: Codable {
  var clangModuleName: String
  var pcmArgs: [String]
  var outputPath: String
}
struct SwiftScanModuleInfo: Codable {
  var swiftModuleName: String
  var outputPath: String
}
```
Whichever makes more sense to deal with in the scanner.

Reply:
ah, That's a great idea. Will differentiate the key for a concise representation.